### PR TITLE
Aho corasick

### LINKED ISSUE #624: Aho-Corasick Algorithm in C++
It is a kind of dictionary-matching algorithm that locates elements of a finite set of strings (the "dictionary") within an input text.

### PR DESCRIPTION
**Pull request by archi14**
Fixes #624 

#### Checklist:
- [x] 4 space indentation.
- [x] Coding conventions are followed.
- [x] Input is taken dynamically.
- [x] Sample Input / Output is added at the end of file.
- [x] Logic Documentation (Comments).
- [x] File names are correct.

#### Changes proposed in this pull request:
- Aho-Corasick Algorithm

#### Languages Used:
- C++

#### Files Added:
- Aho-Corasick c++ code

### CODE REVIEW

File: Aho-Corasick/Aho-Corasick.cpp [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+
+    for (int i = 0; i < words.size(); i++)
+    {   
+        const string &keyword = words[i];
+        int currentState = 0;
+        for (int j = 0; j < keyword.size(); j++)
+        {   
+	    int c = keyword[j] - lowestChar;
+            if (g[currentState][c] == -1)
+            { 	// Allocate a new node
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Put comment in newline.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+            	g[currentState][c] = states++;
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Fix indentation.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+    }
+    // State 0 should have an outgoing edge for all characters.
+    for (int c = 0; c < MAXCHAR; c++)
+    {		
+        if (g[0][c] == -1)
+        {				
+            g[0][c] = 0;	
+        }	
+    }
+	//building the failure function
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Fix indentation.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+    queue<int> q;
+	
+    for (int c = 0; c <= highestChar - lowestChar; c++)
+    { // Iterate over every possible input. All nodes s of depth 1 have fail[s] = 0
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Put the comment in newline.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+        if (g[0][c] != -1 and g[0][c] != 0)
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Use AND operator here (``&&``).

File: Aho-Corasick/Aho-Corasick.cpp [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+        {
+            fail[g[0][c]] = 0;
+            q.push(g[0][c]);	
+        }
+    }
+
+    while (q.size())
+    {        
+	int state = q.front();
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Fix indentation.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+        q.pop();
+        for (int c = 0; c <= highestChar - lowestChar; c++)
+        {
+            if (g[state][c] != -1)
+            {  
+		int failure = fail[state];
+		int failure = fail[state];
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Fix indentation in above two lines.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+    }
+    return g[answer][c];
+}
+
+int main()
+{	
+    vector<string> keywords;
+    cout<<"Enter the number of keywords you want to enter";
+    int n;
+    for(int i = 0; i < n; i++)
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Add space after ``for``.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+    {
+	string temp;
+	cin >> temp;
+	keywords.push_back(temp);
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Fix indentation in above 3 lines.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+    }	
+    cout << "Enter text";
+    string text;
+    cin >> text;
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: No need to add space input (``>>``) and output (``<<``) operators in cin and cout statements.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    buildMachine(keywords, 'a', 'z');
+    int currentState = 0;
+    for (int i = 0; i < text.size(); i++)
+    {
+        currentState = findNextState(currentState, text[i], 'a');
+        if (out[currentState] == 0)
+	{
+	    continue; // Nothing new, moving on to the next character.
+	}
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Fix indentation in above 3 lines.

File: Aho-Corasick/Aho-Corasick.cpp [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+        for (int j = 0; j < keywords.size(); j++)
+        {
+            if (out[currentState] & (1 << j))
+            { // Matched keywords[j]
```
Comment (marwahmanbir) [inline Aho-Corasick/Aho-Corasick.cpp]: Put comment in newline.

**Comment by archi14** [comment]
@marwahmanbir made the changes.

**Pull request merged**